operators/src/concurrency/thread_pool.rs

```diff
@@ -0,0 +1,339 @@
+                        // Or try stealing a task from one of the other threads.
+                        .or_else(|| stealers.iter().map(Stealer::steal).collect())
+                })
+                // Loop while no task was stolen and any steal operation needs to be retried.
+                .find(|s| !s.is_retry())
+                // Extract the stolen task, if there is one.
+                .and_then(Steal::success)
+            });
+
+            if let Some(task) = task {
+                // TODO: recover panics
+                task.call_once();
+                backoff.reset();
+            } else {
+                // TODO: sleep instead of busy waiting?
+                backoff.snooze();
+            }
+        }
+    }
+
+    pub fn create_context(&self) -> ThreadPoolContext {
+        ThreadPoolContext::new(self, self.next_group_id.fetch_add(1))
+    }
+
+    fn compute(&self, task: Task) {
+        self.global_queue.push(task);
+    }
+
+    // TODO: scope for threads
+}
+
+/// A computation context for a group that spawns tasks in a `ThreadPool`
+#[derive(Copy, Clone, Debug)]
+pub struct ThreadPoolContext<'pool> {
+    thread_pool: &'pool ThreadPool,
+    task_group_id: TaskGroupId,
+}
+
+impl<'pool> ThreadPoolContext<'pool> {
+    /// Create a new `ThreadPoolContext`
+    fn new(thread_pool: &'pool ThreadPool, task_group_id: TaskGroupId) -> Self {
+        Self {
+            thread_pool,
+            task_group_id,
+        }
+    }
+
+    /// What is the degree of paralellism that the `ThreadPool` provides?
+    /// This is helpful to determine how to split the work into tasks.
+    pub fn degree_of_parallelism(&self) -> usize {
+        self.thread_pool.degree_of_parallelism
+    }
+
+    /// Compute a task in the `ThreadPool`
+    pub fn compute<F>(&self, task: F)
+    where
+        F: StaticTaskFn,
+    {
+        self.thread_pool
+            .compute(Task::new(self.task_group_id, task));
+    }
+
+    /// Execute a bunch of tasks in a scope that blocks until all tasks are finished.
+    /// Provides a lifetime for that scope.
+    /// TODO: provide an async version so that async workflows can do something in the meantime?
+    /// TODO: handle panics: if a thread panics, this function will block forever
+    pub fn scope<'scope, S>(&'pool self, scope_fn: S)
+    where
+        S: FnOnce(&Scope) + 'scope,
+    {
+        let scope = Scope::<'pool, 'scope> {
+            thread_pool_context: &self,
+            wait_group: WaitGroup::new(),
+            _scope_marker: PhantomData,
+        };
+
+        scope_fn(&scope);
+
+        scope.wait_group.wait();
+    }
+}
+
+/// A scope in which you can execute tasks and it blocks until all tasks are finished
+#[derive(Debug)]
+pub struct Scope<'pool, 'scope> {
+    thread_pool_context: &'pool ThreadPoolContext<'pool>,
+    wait_group: WaitGroup,
+    _scope_marker: PhantomData<&'scope ()>,
+}
+
+impl<'pool, 'scope> Scope<'pool, 'scope> {
```

Comment:
what is this magic? did you invent this, or where is this scope idea and the particular implementation coming from? can't we also wait for specific group id to be finished?

Reply:
The idea is taken from https://docs.rs/crossbeam/0.8.0/crossbeam/fn.scope.html and is incredibly useful because of non-static lifetimes. I did a simple version than the one in `crossbeam` for now.

We could wait for the whole group, but that won't help you with lifetimes inside an operator since groups span over workflow graphs.